tuf/data/keys.go

```diff
@@ -317,113 +327,226 @@ func NewED25519PublicKey(public []byte) *ED25519PublicKey {
-	}
+		privateKey:       privateKey{private: private},
+		signer:           signer{Signer: rsaPrivKey},
+	}, nil
 }
 
 // NewED25519PrivateKey initialized a new ED25519 private key
-func NewED25519PrivateKey(public ED25519PublicKey, private []byte) *ED25519PrivateKey {
+func NewED25519PrivateKey(public ED25519PublicKey, private []byte) (*ED25519PrivateKey, error) {
 	return &ED25519PrivateKey{
 		ED25519PublicKey: public,
-		private:          private,
-	}
+		privateKey:       privateKey{private: private},
+	}, nil
 }
 
 // Private return the serialized private bytes of the key
-func (k ECDSAPrivateKey) Private() []byte {
+func (k privateKey) Private() []byte {
 	return k.private
 }
 
-// Private return the serialized private bytes of the key
-func (k ECDSAx509PrivateKey) Private() []byte {
-	return k.private
+// CryptoSigner returns the underlying crypto.Signer for use cases where we need the default
+// signature or public key functionality (like when we generate certificates)
+func (s signer) CryptoSigner() crypto.Signer {
+	return s.Signer
 }
 
-// Private return the serialized private bytes of the key
-func (k RSAPrivateKey) Private() []byte {
-	return k.private
+// CryptoSigner returns the ED25519PrivateKey which already implements crypto.Signer
+func (k ED25519PrivateKey) CryptoSigner() crypto.Signer {
+	return nil
 }
 
-// Private return the serialized private bytes of the key
-func (k RSAx509PrivateKey) Private() []byte {
-	return k.private
+// CryptoSigner returns the UnknownPrivateKey which already implements crypto.Signer
+func (k UnknownPrivateKey) CryptoSigner() crypto.Signer {
+	return nil
 }
 
-// Private return the serialized private bytes of the key
-func (k ED25519PrivateKey) Private() []byte {
-	return k.private
+type ecdsaSig struct {
+	R *big.Int
+	S *big.Int
 }
 
-// Private return the serialized private bytes of the key
-func (k UnknownPrivateKey) Private() []byte {
-	return k.private
+// Sign creates an ecdsa signature
+func (k ECDSAPrivateKey) Sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ecdsaPrivKey, ok := k.Signer.(*ecdsa.PrivateKey)
+	if !ok {
+		return nil, errors.New("Signer was based on the wrong key type")
+	}
+	hashed := sha256.Sum256(msg)
+	sigASN1, err := k.Signer.Sign(rand, hashed[:], opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertECDSASig(sigASN1, ecdsaPrivKey.Params().BitSize)
+}
+
+// Sign creates an ecdsa signature
+func (k ECDSAx509PrivateKey) Sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ecdsaPrivKey, ok := k.Signer.(*ecdsa.PrivateKey)
+	if !ok {
+		return nil, errors.New("Signer was based on the wrong key type")
+	}
+	hashed := sha256.Sum256(msg)
+	sigASN1, err := k.Signer.Sign(rand, hashed[:], opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertECDSASig(sigASN1, ecdsaPrivKey.Params().BitSize)
+}
+
+func convertECDSASig(sigASN1 []byte, bitSize int) ([]byte, error) {
```

Comment:
It's possible that this format conforms to that spec, I'm actually not sure. This is the way Diogo/Aaron implemented serialization of ECDSA signatures, which I believe was copied from docker/libtrust.